### PR TITLE
AD5758 driver

### PR DESCRIPTION
**Pull request by APorumb**
-updated and restyled AD5758 driver to work with new drivers (driver taken from legacy)

-created AD5758-sdz demo project

### CODE REVIEW

**Review by amiclaus — COMMENTED** [review]
some minor comments on my side.

**Comment by APorumb** [comment]
V2:

-remarks made by @amiclaus solved

**Review by amiclaus — COMMENTED** [review]
I think commits https://github.com/analogdevicesinc/no-OS/pull/1157/commits/604b28c00d2ae2ee35334d3b6830b8fd79eeb112 and https://github.com/analogdevicesinc/no-OS/pull/1157/commits/39cf9dbe091be723a7ce5727805f2b67d65f90bf should be squashed since they have the same purpose: updating the driver.

File: drivers/dac/ad5758/ad5758.c [hunk 1]
```diff
@@ -817,4 +817,4 @@ int32_t ad5758_init(struct ad5758_dev **device,
 	mdelay(1000);
 
 	return ret;
-}
+}
```
Comment (mchindri) [inline drivers/dac/ad5758/ad5758.c]: Change not needed. New line at the end of the file is needed

Comment (mchindri) [reply drivers/dac/ad5758/ad5758.c]: Same next one.

Comment (mchindri) [reply drivers/dac/ad5758/ad5758.c]: In commit description use imperative verbs.
"Move AD5758 driver from legacy"

File: drivers/dac/ad5758/ad5758.h [hunk 2]
```diff
@@ -371,15 +374,25 @@ enum ad5758_adc_mode {
...
+
 struct ad5758_dev {
 	/* SPI */
 	struct spi_desc *spi_desc;
 	/* GPIO */
 	struct gpio_desc *reset_n;
 	struct gpio_desc *ldac_n;
 	/* Device Settings */
-	uint8_t crc_en;
+	enum ad5758_crc_en crc_en;
```
Comment (mchindri) [inline drivers/dac/ad5758/ad5758.h:389]: There is no need to create an enum for this.

File: drivers/dac/ad5758/ad5758.h [hunk 3]
```diff
@@ -401,30 +414,23 @@ struct ad5758_init_param {
...
+int32_t ad5758_set_dc_dc_ilimit(struct ad5758_dev *dev);
 int32_t ad5758_fault_prot_switch_en(struct ad5758_dev *dev, uint8_t enable);
 int32_t ad5758_internal_buffers_en(struct ad5758_dev *dev, uint8_t enable);
-int32_t ad5758_set_out_range(struct ad5758_dev *dev,
-			     enum ad5758_output_range range);
-int32_t ad5758_slew_rate_config(struct ad5758_dev *dev,
-				enum ad5758_slew_rate_clk clk,
-				uint8_t enable);
+int32_t ad5758_set_out_range(struct ad5758_dev *dev);
+int32_t ad5758_slew_rate_config(struct ad5758_dev *dev, uint8_t enable);
```
Comment (mchindri) [inline drivers/dac/ad5758/ad5758.h:426]: Why did you do this?
If use want to change the out_range for example, with this change it can't do that? what is the scope of the function then.
And no, a user shouldn't update struct ad5758_dev.

Note: Commit descriptions should include the why of the change.

File: drivers/dac/ad5758/ad5758.c [hunk 4]
```diff
@@ -102,8 +102,8 @@ static int32_t ad5758_spi_reg_read(struct ad5758_dev *dev,
 		buf[3] = 0x00;
 
 	ret = spi_write_and_read(dev->spi_desc, buf, 4);
-	if (ret < 0)
-		return FAILURE;
+	if (ret)
```
Comment (mchindri) [inline drivers/dac/ad5758/ad5758.c:105]: The check of ret < 0 is good.
A function may return positive values (maybe not in this case, but in general) which is not an error.

File: projects/ad5758-sdz/src/app/ad5758_sdz.c [hunk 5]
```diff
@@ -1,9 +1,9 @@
 /***************************************************************************//**
-* @file ad5758_sdz.c
+* @file ad5758-sdz/src/app/ad5758_sdz.c
```
Comment (mchindri) [inline projects/ad5758-sdz/src/app/ad5758_sdz.c:2]: I would rename this commit to move project from legacy to projects

File: projects/ad5758-sdz/src/app/ad5758_sdz.c [hunk 6]
```diff
@@ -1,9 +1,9 @@
...
 * @brief Implementation of Main Function.
-* @author SPopa (stefan.popa@analog.com)
+* @author Andrei Porumb (andrei.porumb@analog.com)
```
Comment (mchindri) [inline projects/ad5758-sdz/src/app/ad5758_sdz.c:4]: Don't remove first author, just add your signoff above

File: projects/ad5758-sdz/src/app/ad5758_sdz.c [hunk 7]
```diff
@@ -49,46 +49,74 @@
...
-	/* Write mid-scale DAC code to the input register */
-	ad5758_dac_input_write(dev, 0x7FFF);
+	ret = ad5758_init(&dev, &ad5758_default_init_param);
+	if(ret)
+		return FAILURE;
 
-	printf("Done\n");
+	/* Write mid-scale DAC code to the input register
+	 * V1_OUT pin will have half the scale (5V) */
+	ret = ad5758_dac_input_write(dev, 0x7FFF);
```
Comment (mchindri) [inline projects/ad5758-sdz/src/app/ad5758_sdz.c:116]: @analogdevicesinc/noos Do we really want to have a project just for this line?

**Review by mchindri — CHANGES_REQUESTED** [review]
Nitpcik:
Don't add - to commit descriptions.

File: drivers/dac/ad5758/ad5758.h (hunk 2 above)
Comment (APorumb) [reply drivers/dac/ad5758/ad5758.h:389]: In this way, the driver's initial_param is more intuitive to config.

File: drivers/dac/ad5758/ad5758.c (hunk 4 above)
Comment (APorumb) [reply drivers/dac/ad5758/ad5758.c:105]: In this case, ret = 0 only if the function returns SUCCESS. All functions return SUCCESS or FAILURE.

Comment (mchindri) [reply drivers/dac/ad5758/ad5758.c:105]: https://github.com/analogdevicesinc/no-OS/blob/1090f222df0ff8c916ebcfe7990098eaa98aee11/include/error.h#L65

File: drivers/dac/ad5758/ad5758.h (hunk 3 above)
Comment (APorumb) [reply drivers/dac/ad5758/ad5758.h]: Undo-ed.

**Comment by APorumb** [comment]
V3:

-edited accordingly to the remarks

**Review by amiclaus — APPROVED** [review]

File: projects/ad5758-sdz/src/app/ad5758_sdz.c (hunk 7 above)
Comment (amiclaus) [reply projects/ad5758-sdz/src/app/ad5758_sdz.c:116]: It's not just this line, it's also the example that we provide to the customers on how to initialize the device (all the initialization structures that need to be passed to the `ad5758_init` function.

File: drivers/dac/ad5758/ad5758.h (hunk 2 above)
Comment (amiclaus) [reply drivers/dac/ad5758/ad5758.h:389]: If we use this approach in other drivers too, then it should be fine like this. Otherwise, I think it can be dropped, as @mchindri  suggested.

Comment (APorumb) [reply drivers/dac/ad5758/ad5758.h]: enum ad5758_crc_en removed.

**Comment by APorumb** [comment]
V4:

enum ad5758_crc_en removed.

File: drivers/dac/ad5758/ad5758.c [hunk 8]
```diff
@@ -112,17 +116,19 @@ static int32_t ad5758_spi_reg_read(struct ad5758_dev *dev,
...
-	}
+	    (ad5758_compute_crc8(buf, 3) != buf[3]))
+		goto error;
 
 	*reg_data = (buf[1] << 8) | buf[2];
 
 	return SUCCESS;
+
+error:
+	pr_err("%s: CRC Error.\n", __func__);
```
Comment (adrimbarean) [inline drivers/dac/ad5758/ad5758.c:130]: It's not necessarily a CRC error. This message should only show in case ad5758_compute_crc8() returns error. Right now it shows even at spi_write_and_read() errors.

File: drivers/dac/ad5758/ad5758.c [hunk 9]
```diff
@@ -285,23 +289,20 @@ int32_t ad5758_set_dc_dc_conv_mode(struct ad5758_dev *dev,
 	 * The ENABLE_PPC_BUFFERS bit must be set prior to enabling PPC current
 	 * mode.
 	 */
-	if (mode == PPC_CURRENT_MODE) {
+	if (dev->dc_dc_mode == PPC_CURRENT_MODE) {
```
Comment (adrimbarean) [inline drivers/dac/ad5758/ad5758.c:292]: Is this change intended? Before the if clause here checked the new mode parameter. Now it checks the mode currently saved in the handler.

File: drivers/dac/ad5758/ad5758.c [hunk 10]
```diff
@@ -310,9 +311,11 @@ int32_t ad5758_set_dc_dc_conv_mode(struct ad5758_dev *dev,
 		ad5758_spi_reg_read(dev, AD5758_REG_DCDC_CONFIG2, &reg_data);
 	} while (reg_data & AD5758_DCDC_CONFIG2_BUSY_3WI_MSK);
 
-	dev->dc_dc_mode = mode;
```
Comment (adrimbarean) [inline drivers/dac/ad5758/ad5758.c:313]: Do we not want to save the new mode in the handler anymore? If we don't we could erase the dc_dc_mode member from the struct.

File: drivers/dac/ad5758/ad5758.h [hunk 11]
```diff
@@ -273,17 +276,17 @@ enum ad5758_dc_dc_mode {
 };
 
 enum ad5758_dig_diag_flags {
-	DIAG_SPI_CRC_ERR,
-	DIAG_SLIPBIT_ERR,
-	DIAG_SCLK_COUNT_ERR,
+	DIAG_SPI_CRC_ERR = 0,
```
Comment (adrimbarean) [inline drivers/dac/ad5758/ad5758.h:279]: nitpick: this is not necessary. Enum will do this itself.

Comment (APorumb) [reply drivers/dac/ad5758/ad5758.h:279]: This is written in order to respect coding standard.

File: drivers/dac/ad5758/ad5758.c [hunk 12]
```diff
@@ -755,66 +741,112 @@ int32_t ad5758_set_adc_mode(struct ad5758_dev *dev,
...
+err:
+	gpio_remove(dev->ldac_n);
+error_gpio_ldac:
+	gpio_remove(dev->reset_n);
+error_init:
+	pr_err("ad5758 could not be initialized\n");
+	free(dev);
 	return ret;
-}
+}
```
[the remark this answers is not in the document]
Comment (mchindri) [reply drivers/dac/ad5758/ad5758.c]: This should be solved in this commit

File: drivers/dac/ad5758/ad5758.c [hunk 13]
```diff
@@ -166,19 +170,19 @@ static int32_t ad5758_spi_write_mask(struct ad5758_dev *dev,
...
 }
 
 /**
  * Enable/disable SPI CRC function.
  * @param dev - The device structure.
- * @param crc_en - enable or disable
- * Accepted values: 0 : disable
- *		    1 : enable
+ * @param crc_en - CRC status
+ * Accepted values: CRC_DIS
```
Comment (mchindri) [inline drivers/dac/ad5758/ad5758.c:184]: This needs to be undo

File: drivers/dac/ad5758/ad5758.c [hunk 14]
```diff
@@ -231,17 +234,19 @@ int32_t ad5758_soft_reset(struct ad5758_dev *dev)
...
 				   AD5758_KEY_CODE_RESET_1);
-	ret |= ad5758_spi_reg_write(dev, AD5758_REG_KEY,
-				    AD5758_KEY_CODE_RESET_2);
-	if (ret < 0) {
-		printf("%s: Failed.\n", __func__);
-
-		return FAILURE;
-	}
+	if (ret < 0)
+		goto error;
```
Comment (mchindri) [inline drivers/dac/ad5758/ad5758.c:238]: styling: new line is needed after single instruction blocks

File: drivers/dac/ad5758/ad5758.c [hunk 15]
```diff
@@ -231,17 +234,19 @@ int32_t ad5758_soft_reset(struct ad5758_dev *dev)
...
+	ret = ad5758_spi_reg_write(dev, AD5758_REG_KEY,
+				   AD5758_KEY_CODE_RESET_2);
+	if (ret < 0)
+		goto error;
 	/* Wait 100 us */
 	usleep(100);
-
```
Comment (mchindri) [inline drivers/dac/ad5758/ad5758.c:243]: styling: usually new line is added before return

File: drivers/dac/ad5758/ad5758.c [hunk 16]
```diff
@@ -310,9 +311,11 @@ int32_t ad5758_set_dc_dc_conv_mode(struct ad5758_dev *dev,
 		ad5758_spi_reg_read(dev, AD5758_REG_DCDC_CONFIG2, &reg_data);
 	} while (reg_data & AD5758_DCDC_CONFIG2_BUSY_3WI_MSK);
 
-	dev->dc_dc_mode = mode;
-
 	return SUCCESS;
+
+error:
+	pr_err("%s: Failed.\n", __func__);
```
Comment (mchindri) [inline drivers/dac/ad5758/ad5758.c:317]: Maybe: `pr_err("%s: Failed with code: "%PRIi32".\n", __func__, ret);`

**Comment by APorumb** [comment]
V5:

Remarks solved.
Reworked init() function to use init_param structure for initialization. Init_param content is now copied to device structure within each called function.

**Review by adrimbarean — APPROVED** [review]
Looks good to me.

**Comment by APorumb** [comment]
V6:

Found one remark unsolved (new line at the end of the file). Fixed now in first commit.

**Review by mchindri — APPROVED** [review]
Please create a builds.json file to check this project in the CI too

**Comment by APorumb** [comment]
> 

Since there is no public hdl made for this project I can't make a .json file.

**Comment by mchindri** [comment]
> > Please create a builds.json file to check this project in the CI too
> 
> Since there is no public hdl made for this project I can't make a .json file.

Just try :) There is no public hdl for other project neither but they have the builds.json file
as name use ad5758_sdz_zed

**Comment by APorumb** [comment]
V7:

.json file added.

File: projects/ad5758-sdz/builds.json [hunk 17]
```diff
@@ -0,0 +1,8 @@
+{
+      "xilinx": {
+        "example": {
+              "flags" : "",
+              "hardware" : ["ad5758_sdz_zed"]
+        }
+      }
+}
```
Comment (mchindri) [inline projects/ad5758-sdz/builds.json:8]: new line

File: drivers/dac/ad5758/ad5758.c [hunk 18]
```diff
@@ -40,10 +40,15 @@
 /******************************************************************************/
 /***************************** Include Files **********************************/
 /******************************************************************************/
+#include "inttypes.h"
```
Comment (cristiantpop) [inline drivers/dac/ad5758/ad5758.c:43]: Order headers alphabetically, maybe

**Comment by APorumb** [comment]
V8:

Includes are now ordered alphabetically.
